Add a generic kind of banking mandate

CSester opened the pull request:
This pull request splits the account_banking_sepa_direct_debit in 2 modules. The first one is a generic banking mandate module (account_banking_mandate) and the second one is the account_banking_sepa_direct_debit where sdd_mandate now inherits from account_banking_mandate. 

The purpose to this refactoring is to be able to use banking mandates in other use cases such as CH-localized direct debit or any other banking process that requires an authorization.

CSester committed:
Refactoring to get a generic mandate

CSester committed:
Add migration field from 0.1 to 0.2

CSester committed:
PEP8 fixes on migration script

CSester commented:
Rebased on current OCA/7.0 branch. Anyone to review ?

pedrobaeza commented:
Please give time to reviewers to review this significal change, and it's also a bad month...

Thanks.

Regards.

CSester committed:
[ADD] Access rights on account.banking.mandate

CSester committed:
[FIX] Syntax for security file

StefanRijnhart commented:
The idea sounds reasonable, and the implementation looks good at first sight. Thanks for including a migration script. Would it not be easier though to rename tables and columns in the pre-script? Also, please check for version. If it is falsy, you will want to return immediately to prevent the script from running upon module installation.
Looks like you can remove the modified access lines from the sepa module as they are now included in the mandate module.

CSester committed:
Add falsy version check in migration file and removed useless security lines

CSester commented:
So, I rebased to integrate OCA/banking#33 and make Travis green. I also added version check in migration file to prevent running on install. Useless security rules (now being in account_banking_mandate) have been removed.

About the migration script: as the account_banking_mandate is prerequist for the sepa_direct_debit module, it will be installed before sepa_direct_debit upgrade. This way, if I try to simply rename the sdd_mandate table to account_banking_mandate, it will fail because account_banking_mandate table already exists.

nbessi commented on account_banking_mandate/model/account_banking_mandate.py:
```diff
@@ -0,0 +1,185 @@
+        for mandate in self.browse(cr, uid, ids):
+            if (mandate.signature_date and
+                    mandate.signature_date >
+                    datetime.today().strftime('%Y-%m-%d')):
```
please use fields.date.context_today() it should manage TZ for you

nbessi commented on account_banking_mandate/model/account_banking_mandate.py:
```diff
@@ -0,0 +1,185 @@
+        if partner_bank_id:
+            partner_bank_read = self.pool['res.partner.bank'].read(
+                cr, uid, partner_bank_id, ['partner_id'])['partner_id']
+            if partner_bank_read:
```
Isn't partner_id a mandatory field ?

nbessi commented on account_banking_mandate/model/account_banking_mandate.py:
```diff
@@ -0,0 +1,185 @@
+    def validate(self, cr, uid, ids, context=None):
+        to_validate_ids = []
+        for mandate in self.browse(cr, uid, ids, context=context):
+            assert mandate.state == 'draft', 'Mandate should be in draft state'
```
Not blocking but I would return an except_orm Exception, it will be more user friendly

nbessi commented on account_banking_mandate/model/account_banking_mandate.py:
```diff
@@ -0,0 +1,185 @@
+            cr, uid, to_cancel_ids, {'state': 'cancel'}, context=context)
+        return True
+
+    def back2draft(self, cr, uid, ids, context=None):
```
A comment would be welcome

nbessi commented on account_banking_mandate/model/partner_bank.py:
```diff
@@ -0,0 +1,33 @@
+    _inherit = 'res.partner.bank'
+
+    _columns = {
+        'mandate_ids': fields.one2many(
```
An help tooltip would be nice

nbessi commented on account_banking_mandate/__openerp__.py:
```diff
@@ -0,0 +1,49 @@
+    'demo': [],
+    'description': '''
+    This module adds a generic model for banking mandates.
+    These mandates can be specialized to fit any banking mandates (such as sepa
```
A more details description and some explainaition on how to use it would be welcome

nbessi commented on account_banking_mandate/view/account_banking_mandate_view.xml:
```diff
@@ -0,0 +1,122 @@
+		<field name="description">Banking Mandate Cancelled</field>
+	</record>
+</data>
+</openerp>
```
missing carriage

nbessi commented:
@StefanRijnhart @CSester It is ok for me with the migration script.
I found some nit-picks to correct but esle it looks good to me.

CSester committed:
Transformed asserts to exceptions, improved docstrings, add mandate tooltip help

CSester commented:
Modified accordingly to @nbessi comments.

A Pylint test seems to have been added in maintainer quality tools, this way Travis is red. Should that be corrected here or in another PR ?

nbessi commented:
:+1:

pedrobaeza commented:
@CSester, if you have the possibility of fixing here pylint problems, better do it.

Regards.

CSester commented:
@pedrobaeza As OCA/banking#38 fixes the pylint issues, I will rebase as soon as changes are merged in OCA/7.0

gurneyalex commented:
@CSester I know that the code you extracted the new module from had no test whatsoever. Still, this refactoring would be a good time to add a few tests and improve the test coverage of this repository.

gurneyalex commented:
@CSester: I could find no trace of your signing the Contributor Licence Agreement of the community association. Could you please read http://odoo-community.org/page/website.cla and send the signed for the the addresse noted on that page?

CSester commented:
@gurneyalex The CLA is on its way. You should get it soon.

CSester committed:
[ADD] mandate tests

CSester commented:
@gurneyalex I just wrote a test for mandates. I got an issue with the sequence number (only on Travis (running the test locally works). Any idea of what goes wrong?
Edit: Ok, the test failed because sepa mandate sequence and generic mandate sequence had exactly the same format.

CSester committed:
[FIX] generic mandate and sepa mandate reference conflict

ecino commented:
:+1:

gurneyalex commented:
:+1:

gurneyalex commented:
@StefanRijnhart is it ok with you, now the changes have been implemented and tests added?

nbessi commented:
:+1:

StefanRijnhart commented:
Yes, absolutely :+1:
